File: sdks/java/ConformanceTest.java
// Copyright 2026 J3nna Technologies, LLC
// SPDX-License-Identifier: Apache-2.0
//
// Wire-conformance test for the J3nna Mesh Java SDK. Reproduces the canonical signing bytes byte-for-byte
// THROUGH THE SDK's Wire and verifies the reference signatures from jip/conformance/vectors.json — so this
// SDK is wire-compatible with the Go reference (and every other SDK). Sign-side coverage (the vectors carry
// no private keys) is the Identity round-trip at the end.
//
//   javac -cp gson.jar *.java && java -cp .:gson.jar ConformanceTest

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;

public class ConformanceTest {
    private static String S(JsonObject i, String k) { return i.get(k).getAsString(); }
    private static long N(JsonObject i, String k) { return i.get(k).getAsLong(); }

    private static Path findVectors() {
        String[] candidates = {
                "../../../jip/conformance/vectors.json",
                "../../../../jip/conformance/vectors.json",
                "jip/conformance/vectors.json",
                System.getProperty("user.home") + "/web-stt-tts/jip/conformance/vectors.json",
        };
        for (String c : candidates) {
            Path p = Path.of(c);
            if (Files.exists(p)) return p;
        }
        throw new RuntimeException("vectors.json not found in any candidate path");
    }

    // Build the SDK Wire signing bytes from a vector's input JsonObject.
    private static byte[] presence(JsonObject i) {
        var caps = new ArrayList<String>();
        i.getAsJsonArray("capabilities").forEach(c -> caps.add(c.getAsString()));
        return Wire.presenceSigningBytes(
                S(i, "protocol"), S(i, "alg"), S(i, "id"), Wire.unhex(S(i, "public_key_hex")),
                S(i, "endpoint"), S(i, "mcp_path"), caps, (int) N(i, "protocol_major"),
                S(i, "grant_id"), N(i, "heartbeat_unix"));
    }

    private static byte[] grant(JsonObject i) {
        var scopes = new ArrayList<String>();
        i.getAsJsonArray("scopes").forEach(c -> scopes.add(c.getAsString()));
        String principal = i.has("principal") ? i.get("principal").getAsString() : "";
        return Wire.grantSigningBytes(
                S(i, "alg"), S(i, "id"), S(i, "subject"), Wire.unhex(S(i, "public_key_hex")),
                N(i, "tier"), scopes, N(i, "issued_at"), N(i, "not_after"), principal);
    }

    private static byte[] callproof(JsonObject i) {
        return Wire.callproofSigningBytes(
                S(i, "alg"), S(i, "node_id"), S(i, "tool"), Wire.unhex(S(i, "args_hash_hex")),
                N(i, "unix_milli"));
    }

    private static byte[] renewal(JsonObject i) {
        return Wire.renewSigningBytes(
                S(i, "alg"), S(i, "grant_id"), S(i, "subject"), Wire.unhex(S(i, "public_key_hex")),
                N(i, "issued_at"));
    }

    private static byte[] crl(JsonObject i) {
        var ids = new ArrayList<String>(i.getAsJsonObject("revoked").keySet());
        return Wire.crlSigningBytes(S(i, "alg"), N(i, "issued_at"), ids);
    }

    public static void main(String[] args) throws Exception {
        var doc = JsonParser.parseString(Files.readString(findVectors())).getAsJsonObject();
        if (!doc.get("protocol").getAsString().equals("JIP/0.1")) {
            throw new RuntimeException("unexpected protocol");
        }
        int count = 0;
        for (var ve : doc.getAsJsonArray("vectors")) {
            var v = ve.getAsJsonObject();
            var name = v.get("name").getAsString();
            var i = v.getAsJsonObject("input");
            byte[] got = switch (name) {
                case "presence-record" -> presence(i);
                case "grant" -> grant(i);
                case "callproof" -> callproof(i);
                case "renewal" -> renewal(i);
                case "crl" -> crl(i);
                default -> throw new RuntimeException("no Java builder for " + name);
            };
            if (!Wire.enhex(got).equals(v.get("signing_bytes_hex").getAsString())) {
                throw new RuntimeException(name + ": signing bytes differ");
            }

            var pub = Wire.ed25519Pub(Wire.unhex(v.get("signer_public_key_hex").getAsString()));
            var sig = Base64.getDecoder().decode(v.get("signature_b64").getAsString());
            if (!Wire.verify(pub, sig, got)) {
                throw new RuntimeException(name + ": signature did not verify");
            }

            if (name.equals("callproof")) {
                var cj = Wire.canonicalArgsJson(i.getAsJsonObject("args"));
                if (!cj.equals(S(i, "args_canonical_json"))) {
                    throw new RuntimeException("args canonical JSON differs from Go");
                }
                var h = Wire.enhex(MessageDigest.getInstance("SHA-256").digest(Wire.U(cj)));
                if (!h.equals(S(i, "args_hash_hex"))) {
                    throw new RuntimeException("args hash differs");
                }
                // Cross-check the typed-Map canonicalizer used by the SDK call path.
                var map = new java.util.HashMap<String, Object>();
                map.put("count", 3);
                map.put("html", "<b>a&b</b>");
                map.put("message", "hello/world");
                map.put("path", "/v1/tools");
                if (!Wire.canonicalArgsJson(map).equals(S(i, "args_canonical_json"))) {
                    throw new RuntimeException("typed-map canonical JSON differs from Go");
                }
            }
            System.out.println("  ok  " + name);
            count++;
        }

        // Sign-side coverage: generate an identity, persist it, reload, sign with the persisted key, verify
        // with the persisted public key — catches a wrong-seed-slice bug that verify-only vectors can't.
        identityRoundTrip();

        System.out.println("PASS: " + count
                + " vectors verified + identity round-trip (Java wire-compatible with the Go reference)");
    }

    private static void identityRoundTrip() throws Exception {
        Path tmp = Files.createTempFile("mesh-id-", ".json");
        Files.delete(tmp); // ensureIdentity must create it fresh
        try {
            Identity created = Identity.ensureIdentity(tmp.toString());
            Identity loaded = Identity.ensureIdentity(tmp.toString());
            if (!created.id.equals(loaded.id)) {
                throw new RuntimeException("identity id changed on reload");
            }
            if (!java.util.Arrays.equals(created.seed, loaded.seed)
                    || !java.util.Arrays.equals(created.publicKey, loaded.publicKey)) {
                throw new RuntimeException("identity key bytes changed on reload");
            }
            byte[] msg = "the quick brown fox".getBytes(java.nio.charset.StandardCharsets.UTF_8);
            byte[] sig = loaded.sign(msg);
            if (!Wire.verify(loaded.publicKeyObj(), sig, msg)) {
                throw new RuntimeException("identity sign/verify round-trip failed");
            }
            // Persisted priv_b64 must be exactly 64 bytes (seed||pub) per the Go format.
            var blob = JsonParser.parseString(Files.readString(tmp)).getAsJsonObject();
            byte[] raw = Base64.getDecoder().decode(blob.get("priv_b64").getAsString());
            if (raw.length != 64) {
                throw new RuntimeException("persisted priv_b64 is not 64 bytes");
            }
            System.out.println("  ok  identity-roundtrip (gen→persist→reload→sign→verify)");
        } finally {
            Files.deleteIfExists(tmp);
        }
    }
}

File: sdks/dart/test/conformance_test.dart
// Copyright 2026 J3nna Technologies, LLC
// SPDX-License-Identifier: Apache-2.0
//
// Wire-conformance test for the Dart SDK. Reproduces the canonical signing bytes byte-for-byte and verifies
// the reference signatures from jip/conformance/vectors.json — exercising the SDK's own lib/src/wire.dart
// (not a copy). Run:
//
//   dart run test/conformance_test.dart        # exits 0 on pass, non-zero on failure
//
// Pure script (no test runner dep) so `dart run` exercises it directly per the build instructions.

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:j3nna_mesh/src/wire.dart' as wire;

List<int> _unhex(String s) =>
    [for (var i = 0; i < s.length; i += 2) int.parse(s.substring(i, i + 2), radix: 16)];
String _enhex(List<int> b) => b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();

// Locate vectors.json relative to this repo regardless of cwd.
String _vectorsPath() {
  for (final p in [
    '../../../jip/conformance/vectors.json', // from sdks/dart
    '/home/j3nna/web-stt-tts/jip/conformance/vectors.json',
  ]) {
    if (File(p).existsSync()) return p;
  }
  // Fallback: walk up from script dir to find jip/conformance/vectors.json.
  var dir = Directory(Platform.script.toFilePath()).parent;
  for (var i = 0; i < 8; i++) {
    final cand = File('${dir.path}/jip/conformance/vectors.json');
    if (cand.existsSync()) return cand.path;
    dir = dir.parent;
  }
  throw StateError('could not locate jip/conformance/vectors.json');
}

Future<void> main() async {
  final doc = jsonDecode(File(_vectorsPath()).readAsStringSync()) as Map;
  if (doc['protocol'] != 'JIP/0.1') {
    throw 'unexpected protocol ${doc['protocol']}';
  }
  var count = 0;
  for (final vAny in doc['vectors']) {
    final v = vAny as Map;
    final name = v['name'] as String;
    final i = v['input'] as Map;
    Uint8List got;
    switch (name) {
      case 'presence-record':
        got = wire.presenceSigningBytes(
          protocol: i['protocol'] as String,
          alg: i['alg'] as String,
          id: i['id'] as String,
          publicKey: _unhex(i['public_key_hex'] as String),
          endpoint: i['endpoint'] as String,
          mcpPath: i['mcp_path'] as String,
          capabilities: (i['capabilities'] as List).map((e) => '$e').toList(),
          protocolMajor: i['protocol_major'] as int,
          grantId: i['grant_id'] as String,
          heartbeatUnix: i['heartbeat_unix'] as int,
        );
        break;
      case 'grant':
        got = wire.grantSigningBytes(
          alg: i['alg'] as String,
          id: i['id'] as String,
          subject: i['subject'] as String,
          publicKey: _unhex(i['public_key_hex'] as String),
          tier: i['tier'] as int,
          scopes: (i['scopes'] as List).map((e) => '$e').toList(),
          issuedAt: i['issued_at'] as int,
          notAfter: i['not_after'] as int,
          principal: (i['principal'] as String?) ?? '',
        );
        break;
      case 'callproof':
        got = wire.callproofSigningBytes(
          alg: i['alg'] as String,
          nodeId: i['node_id'] as String,
          tool: i['tool'] as String,
          argsHash: _unhex(i['args_hash_hex'] as String),
          unixMilli: i['unix_milli'] as int,
        );
        break;
      case 'renewal':
        got = wire.renewSigningBytes(
          alg: i['alg'] as String,
          grantId: i['grant_id'] as String,
          subject: i['subject'] as String,
          publicKey: _unhex(i['public_key_hex'] as String),
          issuedAt: i['issued_at'] as int,
        );
        break;
      case 'crl':
        got = wire.crlSigningBytes(
          alg: i['alg'] as String,
          issuedAt: i['issued_at'] as int,
          revokedIds: (i['revoked'] as Map).keys.map((e) => '$e').toList(),
        );
        break;
      default:
        throw 'no Dart builder for $name';
    }

    if (_enhex(got) != v['signing_bytes_hex']) {
      throw '$name: signing bytes differ\n  got ${_enhex(got)}\n  exp ${v['signing_bytes_hex']}';
    }

    // ed25519-verify the reference signature over our bytes, via the SDK's verify().
    final pub = _unhex(v['signer_public_key_hex'] as String);
    final sig = base64.decode(v['signature_b64'] as String);
    if (!await wire.verify(pub, sig, got)) {
      throw '$name: signature did not verify';
    }

    if (name == 'callproof') {
      final cj = wire.canonicalArgsJson(i['args'] as Map);
      if (cj != i['args_canonical_json']) {
        throw 'args canonical JSON differs from Go\n  got $cj\n  exp ${i['args_canonical_json']}';
      }
      final h = await wire.argsHash(i['args'] as Map);
      if (_enhex(h) != i['args_hash_hex']) {
        throw 'args hash differs';
      }
    }
    stdout.writeln('  ok  $name');
    count++;
  }
  stdout.writeln('PASS: $count vectors verified (Dart SDK wire-compatible with the Go reference)');
}
